### PR TITLE
[Feature] new request strategy to verify password

### PR DESCRIPTION
**Pull request by David-Henner**
## What's new in this PR?

Addition of a new request strategy to verify the account password.
Since there isn't an endpoint to verify the password, It will use the `update password` endpoint (done similarly on Android).

### CODE REVIEW

**Comment by zenkins** [comment]
Can one of the admins verify this patch?

**Comment by typfel** [comment]
How is this the supposed to be used by the UI project? We usually don't expose request strategies publicly and instead provide an API on the UserSession or any of the model objects.

**Comment by David-Henner** [comment]
@typfel good to know

This is how I used it
```
// On the UI side
// 1. Register to notification
VerifyPasswordStrategy.addPasswordVerifiedObserver(observer:selector:context:)
// 2. Notify RS we want to verify password
VerifyPasswordStrategy.triggerPasswordVerification(password:context:)

// RS notifies when password has been verified
VerifyPasswordStrategy.notifyPasswordVerified(result:context:)
```

File: Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    
+    public static func triggerPasswordVerification(with password: String, context moc: NSManagedObjectContext) {
+        NotificationCenter.default.post(name: .verifyPassword, object: moc, userInfo: [passwordKey: password])
+    }
+    
+    public static func addPasswordVerifiedObserver(_ observer: Any, selector: Selector, context moc: NSManagedObjectContext) {
+        NotificationCenter.default.addObserver(observer, selector: selector, name: .passwordVerified, object: moc)
+    }
+    
+    public static func notifyPasswordVerified(with result: VerifyPasswordResult, context moc: NSManagedObjectContext) {
```
Comment (typfel) [inline Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: Could be private or internal?

File: Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        NotificationCenter.default.post(name: .passwordVerified, object: moc, userInfo: [VerifyPasswordRequestStrategy.verificationResultKey: result])
+    }
+    
+    public static let verificationResultKey = "verificationResultKey"
```
Comment (typfel) [inline Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: Could be private or internal?

Comment (David-Henner) [reply Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: updated in 05705e9

File: Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift (hunk 1 above)
Comment (David-Henner) [reply Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: updated in 05705e9

File: Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    
+    public override func nextRequestIfAllowed() -> ZMTransportRequest? {
+        guard password != nil else {
+            return nil
+        }
+        return self.requestSync.nextRequest()
+    }
+    
+    static func triggerPasswordVerification(with password: String, completion: @escaping (VerifyPasswordResult?) -> Void, context moc: NSManagedObjectContext) {
+        let notificationCenter = NotificationCenter.default
```
Comment (typfel) [inline Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: We use a wrapper around the notification center called `NotificationInContext`, which keeps notifications from multiple user session running in parallel from interfering with each other.

See `ClientUpdateNotification.swift` for an example of how it's used.

Comment (David-Henner) [reply Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: done in 68ed1fc

File: Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    @objc public override init(withManagedObjectContext moc: NSManagedObjectContext, applicationStatus: ApplicationStatus) {
+        self.moc = moc
+        super.init(withManagedObjectContext: moc, applicationStatus: applicationStatus)
+        self.requestSync = ZMSingleRequestSync(singleRequestTranscoder: self, groupQueue: moc)
+        
+        _ = NotificationInContext.addObserver(
+            name: .verifyPassword,
+            context: moc.notificationContext,
+            queue: .main,
```
Comment (typfel) [inline Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: This calls `preparePasswordVerification` which operates on the `VerifyPasswordRequestStrategy` itself which should run on the sync context.

Comment (David-Henner) [reply Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: Right, nice catch

Comment (David-Henner) [reply Source/Synchronization/Strategies/VerifyPasswordRequestStrategy.swift]: Updated 62ecf7b

**Comment by David-Henner** [comment]
@zenkins release minor

**Comment by David-Henner** [comment]
@zenkins release minor

**Comment by David-Henner** [comment]
@zenkins release minor

**Pull request merged**